### PR TITLE
win,tty: bring back and fix ANSI support

### PR DESCRIPTION
**Pull request by santigimeno**
Two commits:
- Revert the revert.
- When calling `MultiByteToWideChar` make sure there's enough room in the output buffer by dynamically allocating memory in case the size of the buffer needs to be greater than 8192 characters.

### CODE REVIEW

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/174/

**Comment by santigimeno** [comment]
Thoughts? @iWuzHere @libuv/collaborators

File: src/win/tty.c [hunk 1]
```diff
@@ -1636,6 +1667,51 @@ static int uv_tty_write_bufs(uv_tty_t* handle,
...
+                                           0);
+
+      if (utf16_buf_used == 0) {
+        *error = GetLastError();
+        break;
+      }
+
+      max_len = (utf16_buf_used + 1) * sizeof(WCHAR);
+      if (max_len > MAX_CONSOLE_CHAR) {
+        utf16_ansi_buf = (WCHAR*)malloc(max_len);
```
Comment (iWuzHere) [inline src/win/tty.c]: Shou;dn't we be using `uv_malloc` and `uv_free`?

File: src/win/tty.c [hunk 2]
```diff
@@ -1636,6 +1667,51 @@ static int uv_tty_write_bufs(uv_tty_t* handle,
...
+        if (!MultiByteToWideChar(CP_UTF8,
```
Comment (iWuzHere) [inline src/win/tty.c]: I think this section can be rewritten so we only have one `MultiByteToWideChar` call

File: src/win/tty.c [hunk 3]
```diff
@@ -1614,6 +1631,20 @@ static int uv_tty_write_bufs(uv_tty_t* handle,
...
+  do {                                                               \
+    pos = 0;                                                         \
+    do {                                                             \
+      len = utf16_buf_used - pos;                                    \
+      if (len > MAX_CONSOLE_CHAR)                                    \
+        len = MAX_CONSOLE_CHAR;                                      \
+      uv_tty_emit_text(handle, &utf16_ansi_buf[pos], len, error);    \
+      pos += len;                                                    \
+    } while (pos < utf16_buf_used);                                  \
+    free(utf16_ansi_buf);                                            \
```
Comment (iWuzHere) [inline src/win/tty.c]: use `uv_free`

**Comment by iWuzHere** [comment]
Thanks to getting to is so quickly! Left some comments. Also great job adding regression tests!

File: src/win/tty.c (hunk 1 above)
Comment (saghul) [inline src/win/tty.c]: Indeed. Also, unneeded cast.

File: test/test-tty.c [hunk 4]
```diff
@@ -212,6 +212,89 @@ TEST_IMPL(tty_raw) {
...
+  ttyout_fd = _open_osfhandle((intptr_t) handle, 0);
+
+  ASSERT(ttyout_fd >= 0);
+
+  ASSERT(UV_TTY == uv_guess_handle(ttyout_fd));
+
+  r = uv_tty_init(uv_default_loop(), &tty_out, ttyout_fd, 0);  /* Writable. */
+  ASSERT(r == 0);
+
+  char dummy[1];
```
Comment (saghul) [inline test/test-tty.c]: move declarations to the top of old VS will complain

**Comment by santigimeno** [comment]
Updated with with suggested changes.

@iWuzHere I have tried make only one `MultiByteToWideChar` call and having only one `FLUSH_TEXT()`. I'm not sure the code is now cleaner than it was when separated. Just let me know what you think. Thanks

File: test/test-tty.c [hunk 5]
```diff
@@ -212,6 +212,89 @@ TEST_IMPL(tty_raw) {
...
+  ttyout_fd = _open_osfhandle((intptr_t) handle, 0);
+
+  ASSERT(ttyout_fd >= 0);
+
+  ASSERT(UV_TTY == uv_guess_handle(ttyout_fd));
+
+  r = uv_tty_init(uv_default_loop(), &tty_out, ttyout_fd, 0);  /* Writable. */
+  ASSERT(r == 0);
+
+  uv_buf_t bufs[1];
```
Comment (saghul) [inline test/test-tty.c]: declare at the top

**Comment by saghul** [comment]
Random note (no need to act on it): I wish we could refactor that not to need the macros, it's pretty damn hard to follow :-S

**Comment by santigimeno** [comment]
Updated

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/180/

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/445e3a1f06b2a049e2b50b7edf2992ce80167014 and https://github.com/libuv/libuv/commit/c2f0e4f64ef219fda1f52e2987430f1bea52231d. Thanks!
